internal/alb/sg/security_group.go

```diff
@@ -0,0 +1,255 @@
+)
+
+// SecurityGroup represents an SecurityGroup resource in AWS
+type SecurityGroup struct {
+	// We identify SecurityGroup either by GroupID or GroupName
+	GroupID   *string
+	GroupName *string
+
+	InboundPermissions []*ec2.IpPermission
+}
+
+// SecurityGroupController manages SecurityGroups
+type SecurityGroupController interface {
+	// Reconcile ensures the securityGroup exists and match the specification.
+	// Field GroupID or GroupName will be populated if unspecified.
+	Reconcile(*SecurityGroup) error
+
+	// Delete ensures the securityGroup does not exists.
+	Delete(*SecurityGroup) error
+}
+
+type securityGroupController struct {
+	ec2    *albec2.EC2
+	logger *log.Logger
+}
+
+func (controller *securityGroupController) Reconcile(group *SecurityGroup) error {
+	instance, err := controller.findExistingSGInstance(group)
+	if err != nil {
+		return err
+	}
+	if instance != nil {
+		return controller.reconcileByModifySGInstance(group, instance)
+	}
+	return controller.reconcileByNewSGInstance(group)
+}
+
+func (controller *securityGroupController) Delete(group *SecurityGroup) error {
+	if group.GroupID != nil {
+		return controller.ec2.DeleteSecurityGroupByID(*group.GroupID)
+	}
+	instance, err := controller.findExistingSGInstance(group)
+	if err != nil {
+		return err
+	}
+	if instance != nil {
+		return controller.ec2.DeleteSecurityGroupByID(*instance.GroupId)
+	}
+	return nil
+}
+
+func (controller *securityGroupController) reconcileByNewSGInstance(group *SecurityGroup) error {
+	// TODO: move these VPC calls into controller startup, and part of controller configuration
+	vpcID, err := controller.ec2.GetVPCID()
+	if err != nil {
+		return err
+	}
+	createSGOutput, err := controller.ec2.CreateSecurityGroup(&ec2.CreateSecurityGroupInput{
+		VpcId:       vpcID,
+		GroupName:   group.GroupName,
+		Description: aws.String("Instance SecurityGroup created by alb-ingress-controller"),
+	})
+	if err != nil {
+		return err
+	}
+	group.GroupID = createSGOutput.GroupId
+
+	_, err = controller.ec2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
+		GroupId:       group.GroupID,
+		IpPermissions: group.InboundPermissions,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = controller.ec2.CreateTags(&ec2.CreateTagsInput{
+		Resources: []*string{group.GroupID},
+		Tags: []*ec2.Tag{
+			{
+				Key:   aws.String("Name"),
+				Value: group.GroupName,
+			},
+			{
+				Key:   aws.String(albec2.ManagedByKey),
+				Value: aws.String(albec2.ManagedByValue),
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (controller *securityGroupController) reconcileByModifySGInstance(group *SecurityGroup, instance *ec2.SecurityGroup) error {
+	if group.GroupID == nil {
+		group.GroupID = instance.GroupId
+	}
+	if group.GroupName == nil {
+		group.GroupName = instance.GroupName
+	}
+
+	permissionsToRevoke := diffIPPermissions(instance.IpPermissions, group.InboundPermissions)
```

Comment:
Can we document what variables mean here? It's not clear to me what the desired state is here (e.g. source is `instance`, and target is `group`). Are we trying to reconcile `instance` so that `group` matches the permissions of `instance`?

Reply:
added method doc to that method, group is our internal securityGroup definition, which contains the desired state. instance is the AWS securityGroup representation